README.md
# mini_ui

Mini UI is a Flutter package. It is a Sample for Flutter UI Design System.

**Step 1:**
Add `mini_ui : ^1.0.0` to your `pubspec.yaml`

**Step 2:**

```dart
import 'package:mini_ui/mini_ui.dart';
```

**Step 3:**

Boom! Play with Mini UI

## Buttons

![Button Usage](assets/imgs/Buttons.png?raw=true "Buttons")

## Text Widgets

![Text Widgets](assets/imgs/Texts.png?raw=true "Text Widgets")

## Input Fields

![Input Fields](assets/imgs/Input_fields.png?raw=true "Input Fields")


CHANGELOG.md
## 1.0.0

- Initial build and Structuring.


test/mini_ui_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:mini_ui/mini_ui.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/shared/styles.dart
import 'package:flutter/material.dart';

/// Heading one TextStyle
const TextStyle minHeading1Style = TextStyle(
  fontSize: 32,
  fontWeight: FontWeight.w700,
  letterSpacing: 0.32,
);

/// Heading 2 TextStyle
const TextStyle minHeading2Style = TextStyle(
  fontSize: 24,
  fontWeight: FontWeight.w600,
  letterSpacing: 0.24,
);

/// SubHeading TextStyle
const TextStyle minSubheadingStyle = TextStyle(
  fontSize: 20,
  fontWeight: FontWeight.w500,
  letterSpacing: 0.20,
);

/// Body one TextStyle
const TextStyle minBody1Style = TextStyle(
  fontSize: 16,
  letterSpacing: 0.16,
);

/// Body two TextStyle
const TextStyle minBody2Style = TextStyle(
  fontSize: 13,
  letterSpacing: 0.13,
);

/// Label TextStyle
const TextStyle minLabelStyle = TextStyle(
  fontSize: 10,
  letterSpacing: 0.10,
);


lib/src/shared/helpers.dart
import 'package:flutter/material.dart';

/// Horizontal spaces using [SizedBox] widgets
const Widget hlSpaceXS = SizedBox(width: 5.0);
const Widget hlSpaceS = SizedBox(width: 10.0);
const Widget hlSpaceM = SizedBox(width: 18.0);
const Widget hlSpaceL = SizedBox(width: 24.0);
const Widget hlSpaceXL = SizedBox(width: 32.0);
const Widget hlSpaceXXL = SizedBox(width: 50.0);

/// Vertical spaces using [SizedBox] widgets
const Widget vlSpaceXS = SizedBox(height: 5.0);
const Widget vlSpaceS = SizedBox(height: 10.0);
const Widget vlSpaceM = SizedBox(height: 18.0);
const Widget vlSpaceL = SizedBox(height: 24.0);
const Widget vlSpaceXL = SizedBox(height: 31.0);
const Widget vlSpaceXXL = SizedBox(height: 50.0);

/// Returns a screen width by using [context]
double screenWidth(BuildContext context) => MediaQuery.of(context).size.width;

/// Returns a screen height by using [context]
double screenHeight(BuildContext context) => MediaQuery.of(context).size.height;

/// Returns a screen height in percentages
double screenHeightPercentage(BuildContext context, {double percentage = 1}) =>
    screenHeight(context) * percentage;

/// Returns a screen width in percentages
double screenWidthPercentage(BuildContext context, {double percentage = 1}) =>
    screenWidth(context) * percentage;


lib/src/widgets/mini_text.dart
import 'package:flutter/material.dart';
import 'package:mini_ui/src/shared/styles.dart';

/// [MiniText] is a single text widget with many constructors
///
/// Each constructors is using different Typography stylings from the styles.
///
/// To make the text center using [align] = [TextAlign.center]
class MiniText extends StatelessWidget {
  final String text;
  final TextStyle style;
  final bool? underline;
  final TextAlign alignment;

  const MiniText.heading1(
    this.text, {
    Key? key,
    TextAlign align = TextAlign.start,
    this.underline = false,
  })  : style = minHeading1Style,
        alignment = align,
        super(key: key);

  const MiniText.heading2(
    this.text, {
    Key? key,
    TextAlign align = TextAlign.start,
    this.underline = false,
  })  : style = minHeading2Style,
        alignment = align,
        super(key: key);

  const MiniText.subHeading(
    this.text, {
    Key? key,
    TextAlign align = TextAlign.start,
    this.underline = false,
  })  : style = minSubheadingStyle,
        alignment = align,
        super(key: key);

  const MiniText.body1(
    this.text, {
    Key? key,
    TextAlign align = TextAlign.start,
    this.underline = false,
  })  : style = minBody1Style,
        alignment = align,
        super(key: key);

  const MiniText.body2(
    this.text, {
    Key? key,
    TextAlign align = TextAlign.start,
    this.underline = false,
  })  : style = minBody2Style,
        alignment = align,
        super(key: key);

  const MiniText.label(
    this.text, {
    Key? key,
    TextAlign align = TextAlign.start,
    this.underline = false,
  })  : style = minLabelStyle,
        alignment = align,
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return Text(
      text,
      style: style.copyWith(
        decoration: underline! ? TextDecoration.underline : TextDecoration.none,
      ),
      textAlign: alignment,
    );
  }
}


lib/src/widgets/mini_button.dart
import 'package:flutter/material.dart';
import 'package:mini_ui/src/shared/colors.dart';
import 'package:mini_ui/src/shared/helpers.dart';
import 'package:mini_ui/src/shared/styles.dart';

/// MiniButton
///
/// [title]  A text which is displayed on the button
///
/// Shows [CircularProgressIndicator] if [busy] = true.
class MiniButton extends StatelessWidget {
  final String title;
  final bool disabled;
  final bool busy;
  final void Function() onTap;
  final bool outline;
  final Widget? prefix;
  final Widget? suffix;

  const MiniButton({
    Key? key,
    required this.title,
    this.disabled = false,
    this.busy = false,
    required this.onTap,
    this.prefix,
    this.suffix,
  })  : outline = false,
        super(key: key);

  const MiniButton.outline({
    Key? key,
    required this.title,
    required this.onTap,
    this.prefix,
    this.suffix,
    this.busy = false,
    this.disabled = false,
  })  : outline = true,
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return MouseRegion(
      cursor: SystemMouseCursors.forbidden,
      child: AbsorbPointer(
        absorbing: disabled || busy,
        child: AnimatedContainer(
          duration: const Duration(milliseconds: 350),
          child: Material(
            color: outline ? miniTransparent : miniPrimaryDark,
            borderRadius: BorderRadius.circular(8),
            child: InkWell(
              onTap: onTap,
              borderRadius: BorderRadius.circular(8),
              splashColor: outline
                  ? miniSecondaryLight
                  : miniSecondaryLight.withOpacity(0.4),
              child: Container(
                width: double.infinity,
                height: 48,
                alignment: Alignment.center,
                decoration: outline
                    ? BoxDecoration(
                        color: Colors.transparent,
                        borderRadius: BorderRadius.circular(8),
                        border: Border.all(
                          color: miniPrimaryDark,
                          width: 1,
                        ),
                      )
                    : BoxDecoration(
                        color: disabled ? miniPrimaryLight : miniTransparent,
                        borderRadius: BorderRadius.circular(8),
                      ),
                child: !busy
                    ? Row(
                        mainAxisSize: MainAxisSize.min,
                        children: [
                          if (prefix != null) prefix!,
                          if (prefix != null) hlSpaceS,
                          Text(
                            title,
                            style: minBody1Style.copyWith(
                              fontWeight:
                                  outline ? FontWeight.w400 : FontWeight.bold,
                              color: outline
                                  ? miniPrimaryDark
                                  : miniBackgroundColor,
                            ),
                          ),
                          if (suffix != null) hlSpaceS,
                          if (suffix != null) suffix!,
                        ],
                      )
                    : SizedBox(
                        height: 30,
                        width: 30,
                        child: CircularProgressIndicator(
                          strokeWidth: 3,
                          valueColor: AlwaysStoppedAnimation(
                            outline ? miniPrimaryDark : Colors.white,
                          ),
                        ),
                      ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}


lib/src/widgets/mini_input_field.dart
import 'package:flutter/material.dart';
import 'package:mini_ui/mini_ui.dart';

/// [MiniInputField] is using the default theme of the package you can customize it.
///
/// Using [password] you can make the text obscure
///
/// It is [TextField] cannot be used inside the [Form] widget
class MiniInputField extends StatelessWidget {
  final TextEditingController controller;
  final String placeholder;
  final Widget? prefix;
  final Widget? suffix;
  final bool password;
  final circularBorder = OutlineInputBorder(
    borderRadius: BorderRadius.circular(8),
  );

  MiniInputField({
    Key? key,
    required this.controller,
    this.placeholder = "",
    this.prefix,
    this.suffix,
    this.password = false,
  }) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Theme(
      data: ThemeData(primaryColor: miniPrimaryDark),
      child: TextField(
        controller: controller,
        // style: const TextStyle(height: 1),
        obscureText: password,
        decoration: InputDecoration(
          hintText: placeholder,
          contentPadding:
              const EdgeInsets.symmetric(vertical: 15, horizontal: 20),
          filled: true,
          fillColor: miniTertiaryLight.withOpacity(0.25),
          prefixIcon: prefix,
          suffixIcon: suffix,
          border: circularBorder.copyWith(
            borderSide: const BorderSide(color: miniTertiaryLight),
          ),
          errorBorder: circularBorder.copyWith(
            borderSide: const BorderSide(color: Colors.red),
          ),
          focusedBorder: circularBorder.copyWith(
            borderSide: const BorderSide(color: miniSecondaryLight),
          ),
          enabledBorder: circularBorder.copyWith(
            borderSide: const BorderSide(color: miniTertiaryLight),
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:mini_ui/mini_ui.dart';

void main() {
  runApp(MyApp());
}

/// Playground and showcase of the **Mini UI**.
///
/// All variety of buttons
///  * Normal button
///  * Loading button
///  * Disabled button
///  * Outline button
///  * Loading outline button
///  * Disabled outline button
///  * Button with suffix Icon
///  * Button with Prefix Icon
///
/// All variety of Text Widgets
///  * Heading 1
///  * Heading 2
///  * Subheading
///  * Body 1
///  * Body 2
///  * Label
///
/// Text Fields
///  * Normal example
///  * Obscure example
///
class MyApp extends StatelessWidget {
  MyApp({Key? key}) : super(key: key);
  bool isPasswordVisible = false;
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: ListView(
          padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 24),
          children: [
            const MiniText.heading1(
              "Mini UI",
              align: TextAlign.center,
              underline: true,
            ),
            ...textWidgets,
            vlSpaceXL,
            ...buttons,
            vlSpaceL,
            ...inputFields,
          ],
        ),
      ),
    );
  }

  ///List of button widgets
  List<Widget> get buttons => [
        MiniButton(
          title: "Hello",
          onTap: () {},
        ),
        vlSpaceL,
        MiniButton(
          title: "Loading",
          onTap: () {},
          busy: true,
        ),
        vlSpaceL,
        MiniButton(
          title: "Disabled",
          onTap: () {},
          // busy: true,
          disabled: true,
        ),
        vlSpaceL,
        MiniButton.outline(
          title: "Outline",
          onTap: () {},
        ),
        vlSpaceL,
        MiniButton.outline(
          title: "Outline",
          busy: true,
          onTap: () {},
        ),
        vlSpaceL,
        MiniButton.outline(
          title: "Disabled",
          disabled: true,
          onTap: () {},
        ),
        vlSpaceL,
        MiniButton(
          title: "Send",
          onTap: () {},
          suffix: const Icon(
            Icons.send,
            color: miniBackgroundColor,
            size: 20,
          ),
        ),
        vlSpaceL,
        MiniButton(
          title: "Download",
          onTap: () {},
          prefix: const Icon(
            Icons.download,
            color: miniBackgroundColor,
            size: 20,
          ),
        ),
        vlSpaceL,
      ];

  /// List of Text Widgets
  List<Widget> get textWidgets => [
        const MiniText.heading1("Heading 1"),
        vlSpaceXS,
        const MiniText.heading2("Heading 2"),
        vlSpaceXS,
        const MiniText.subHeading("Subheading"),
        vlSpaceXS,
        const MiniText.body1("Body 1"),
        vlSpaceXS,
        const MiniText.body2("Body 2"),
        vlSpaceXS,
        const MiniText.label("Label"),
      ];

  /// List of TextFields
  List<Widget> get inputFields => [
        MiniInputField(
          controller: TextEditingController(),
          placeholder: "Email",
          prefix: const Icon(
            Icons.email,
            color: miniPrimaryDark,
          ),
          suffix: IconButton(
            splashRadius: 18,
            iconSize: 20,
            onPressed: () {},
            icon: const Icon(
              Icons.clear,
              color: miniPrimaryDark,
            ),
          ),
        ),
        vlSpaceS,
        MiniInputField(
          controller: TextEditingController(),
          placeholder: "Password",
          prefix: const Icon(
            Icons.lock,
            color: miniPrimaryDark,
          ),
          suffix: IconButton(
            splashRadius: 18,
            iconSize: 20,
            onPressed: () {
              isPasswordVisible = !isPasswordVisible;
            },
            icon: Icon(
              isPasswordVisible ? Icons.remove_red_eye : Icons.visibility_off,
              color: miniPrimaryDark,
            ),
          ),
          password: isPasswordVisible,
        )
      ];
}
